### PR TITLE
Libraries authentication with testing

### PR DESCRIPTION
**Pull request by Denniswillie**
Add several changes:

1. Add web.xml to map servlets and welcome file
2. Restructure LoadRecommendationsServlet.java so that if user is not logged in, server will not read from database
3. Restructure CreateRecommendationServlet.java so that is the user is not logged in, user will be prompted to a login page, and after that, will be redirected to "/recommendations.html"
4. Restructure script.js so that it has 2 additional functions, to display recommendation page and the other to display login page (link to login url) 
5. Add dependencies on pom.xml
6. Add CreateRecommendationServletTest.java and LoadRecommendationsServletTest.java

(note: still trying to figure out how to pass query string parameters to the get request on the unit test on LoadRecommendationsServletTest.java)

### CODE REVIEW

**Comment by cgollner** [comment]
> Add several changes:
> 
> 1. Add web.xml to map servlets and welcome file
> 2. Restructure LoadRecommendationsServlet.java so that if user is not logged in, server will not read from database
> 3. Restructure CreateRecommendationServlet.java so that is the user is not logged in, user will be prompted to a login page, and after that, will be redirected to "/recommendations.html"
> 4. Restructure script.js so that it has 2 additional functions, to display recommendation page and the other to display login page (link to login url)
> 5. Add dependencies on pom.xml
> 6. Add CreateRecommendationServletTest.java and LoadRecommendationsServletTest.java
> 
> (note: still trying to figure out how to pass query string parameters to the get request on the unit test on LoadRecommendationsServletTest.java)

Nice PR description!

File: portfolio/src/main/java/com/google/sps/servlets/CreateRecommendationServlet.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.appengine.api.users.User;
+
+@WebServlet("/create-recommendation")
+public class CreateRecommendationServlet extends HttpServlet {
+
+  public final UserService userService;
```
Comment (cgollner) [inline portfolio/src/main/java/com/google/sps/servlets/CreateRecommendationServlet.java:40]: This field should be private

File: portfolio/src/main/java/com/google/sps/servlets/LoadRecommendationsServlet.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+            Recommendation recommendation = new Recommendation(id, name, relationship, comment, email);
+            recommendations.add(recommendation);
+        }
+
+        RecommendationsResponse recommendationsResponse = new RecommendationsResponse(recommendations, maxNumberOfRecommendations, true, userService.createLogoutURL("/recommendations.html"));
+        response.getWriter().println(gson.toJson(recommendationsResponse)); 
+    }
+  }
+
+  public int queryStringParser(String queryString){
```
Comment (cgollner) [inline portfolio/src/main/java/com/google/sps/servlets/LoadRecommendationsServlet.java:99]: Let's remove this method and use the official one from HttpServletRequest

File: portfolio/src/main/webapp/script.js [hunk 3]
```diff
@@ -13,34 +27,49 @@ function loadRecommendations(queryString) {
     
     const scrollElement = document.querySelector(".specialScroll");
     
+    //if user is logged in
```
Comment (cgollner) [inline portfolio/src/main/webapp/script.js:30]: Remove code comment. It doesn't tell us more than the condition in the line below does.

File: portfolio/src/test/java/com/google/sps/CreateRecommendationServletTest.java [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+        helper.setUp();
+    }
+
+    //test if datastore has correct data when user is logged in
+    @Test
+    public void dataStoreHasCorrectDataIfLoggedInTest() throws IOException {
+
+        login("denniswillie", "google.com", true);
+
+        // create mock objects
```
Comment (cgollner) [inline portfolio/src/test/java/com/google/sps/CreateRecommendationServletTest.java:62]: Usually we do the creation of mocks in the `setUp` method above, and store them as fields.

File: portfolio/src/test/java/com/google/sps/CreateRecommendationServletTest.java [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+        DatastoreService dataStoreService = DatastoreServiceFactory.getDatastoreService();
+        UserService userService = UserServiceFactory.getUserService();
+
+        //stubbing
+        when(request.getParameter("name")).thenReturn("Dennis");
+        when(request.getParameter("relationship")).thenReturn("Myself");
+        when(request.getParameter("comment")).thenReturn("I have very tiny legs");
+
+        //create servlet and call doPost method
+        CreateRecommendationServlet createRecommendationServlet = new CreateRecommendationServlet(userService, dataStoreService);
```
Comment (cgollner) [inline portfolio/src/test/java/com/google/sps/CreateRecommendationServletTest.java:76]: The servlet could also be created in the "setUp" method.

File: portfolio/src/test/java/com/google/sps/CreateRecommendationServletTest.java [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+        createRecommendationServlet.doPost(request, response);
+
+        //create and send query to database
+        Query testQuery = new Query("Recommendation");
+        PreparedQuery testResult = dataStoreService.prepare(testQuery);
+        
+        //assert that the datastore has 1 entity with the kind name = "Recommendation" with all the properties
+        Gson gson = new Gson();
+        String testObject = gson.toJson(testResult.asSingleEntity());
```
Comment (cgollner) [inline portfolio/src/test/java/com/google/sps/CreateRecommendationServletTest.java:85]: No need to convert this to json. We can just compare the java object directly.

File: portfolio/src/test/java/com/google/sps/CreateRecommendationServletTest.java [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+    }
+
+    @Before
+    public void setUp() {
+        helper.setUp();
+    }
+
+    //test if datastore has correct data when user is logged in
+    @Test
+    public void dataStoreHasCorrectDataIfLoggedInTest() throws IOException {
```
Comment (cgollner) [inline portfolio/src/test/java/com/google/sps/CreateRecommendationServletTest.java:58]: Good name for the test, but the convention is slightly different. 

One common pattern of test names is as follows:
```
<methodUnderTest>_<condition>_<expectedOutcome>
<behaviorAcrossManyMethods>_<condition1>_<condition2>_<expectedOutcome>
```
A test's name should fully describe what a test does, including not only the state and details of the method being tested but also its expected outcome.

File: portfolio/src/test/java/com/google/sps/CreateRecommendationServletTest.java [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.FetchOptions.Builder;
+import com.google.appengine.api.datastore.FetchOptions;
+
+@RunWith(JUnit4.class)
+public class CreateRecommendationServletTest extends CreateRecommendationServlet{
```
Comment (cgollner) [inline portfolio/src/test/java/com/google/sps/CreateRecommendationServletTest.java:40]: Awesome that you are starting to write tests!

File: portfolio/src/main/java/com/google/sps/servlets/CreateRecommendationServlet.java (hunk 1 above)
Comment (Denniswillie) [reply portfolio/src/main/java/com/google/sps/servlets/CreateRecommendationServlet.java]: changed to private final

File: portfolio/src/main/java/com/google/sps/servlets/LoadRecommendationsServlet.java (hunk 2 above)
Comment (Denniswillie) [reply portfolio/src/main/java/com/google/sps/servlets/LoadRecommendationsServlet.java]: done

File: portfolio/src/main/webapp/script.js (hunk 3 above)
Comment (Denniswillie) [reply portfolio/src/main/webapp/script.js]: done

File: portfolio/src/test/java/com/google/sps/CreateRecommendationServletTest.java (hunk 4 above)
Comment (Denniswillie) [reply portfolio/src/test/java/com/google/sps/CreateRecommendationServletTest.java]: done

File: portfolio/src/test/java/com/google/sps/CreateRecommendationServletTest.java (hunk 5 above)
Comment (Denniswillie) [reply portfolio/src/test/java/com/google/sps/CreateRecommendationServletTest.java]: done

File: portfolio/src/test/java/com/google/sps/CreateRecommendationServletTest.java (hunk 6 above)
Comment (Denniswillie) [reply portfolio/src/test/java/com/google/sps/CreateRecommendationServletTest.java]: okay

File: portfolio/src/test/java/com/google/sps/CreateRecommendationServletTest.java (hunk 7 above)
Comment (Denniswillie) [reply portfolio/src/test/java/com/google/sps/CreateRecommendationServletTest.java]: changed.

File: portfolio/src/test/java/com/google/sps/CreateRecommendationServletTest.java (hunk 8 above)
Comment (Denniswillie) [reply portfolio/src/test/java/com/google/sps/CreateRecommendationServletTest.java:41]: thanks!

File: portfolio/src/test/java/com/google/sps/CreateRecommendationServletTest.java [hunk 9]
```diff
@@ -51,64 +59,41 @@ public void login(String username, String domain, boolean isAdmin) {
...
-        Gson gson = new Gson();
-        String testObject = gson.toJson(testResult.asSingleEntity());
-        Assert.assertTrue(testObject.contains("\"name\":\"Dennis\"")&&
-                        testObject.contains("\"comment\":\"I have very tiny legs\"")&&
-                        testObject.contains("\"relationship\":\"Myself\"")&&
-                        testObject.contains("\"email\":\"denniswillie@google.com\""));
+        Assert.assertTrue(testResult.getProperty("name").equals("Dennis") &&
+                        testResult.getProperty("relationship").equals("Myself") &&
+                        testResult.getProperty("comment").equals("I have very tiny legs") &&
+                        testResult.getProperty("email").equals("denniswillie@google.com"));
```
Comment (cgollner) [inline portfolio/src/test/java/com/google/sps/CreateRecommendationServletTest.java:90]: We usually use `assertEquals()` to make it easier to understand what failed. If this fails now, you will only see that something wasn't equal, but won't be able to tell what it was.

**Pull request merged**